actors/builtin/market/market_actor.go

```diff
@@ -289,65 +289,118 @@ func (A Actor) VerifyDealsForActivation(rt Runtime, params *VerifyDealsForActiva
 	}
 }
 
-type ActivateDealsParams struct {
+type ActivateSectorRequest struct {
+	SectorNumber abi.SectorNumber
 	DealIDs      []abi.DealID
 	SectorExpiry abi.ChainEpoch
 }
 
+type ActivateDealsParam struct {
+	Requests []ActivateSectorRequest
+}
+
+type ActivateDealsReturn struct {
+	ActivatedSectors []abi.SectorNumber
+}
+
 // Verify that a given set of storage deals is valid for a sector currently being ProveCommitted,
 // update the market's internal state accordingly.
-func (a Actor) ActivateDeals(rt Runtime, params *ActivateDealsParams) *adt.EmptyValue {
+func (a Actor) ActivateDeals(rt Runtime, params *ActivateDealsParam) *ActivateDealsReturn {
 	rt.ValidateImmediateCallerType(builtin.StorageMinerActorCodeID)
 	minerAddr := rt.Message().Caller()
 	currEpoch := rt.CurrEpoch()
 
 	var st State
 	store := adt.AsStore(rt)
 
+	activatedSectors := make([]abi.SectorNumber, 0, len(params.Requests))
+
 	// Update deal dealStates.
 	rt.State().Transaction(&st, func() {
-		_, _, err := ValidateDealsForActivation(&st, store, params.DealIDs, minerAddr, params.SectorExpiry, currEpoch)
-		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to validate dealProposals for activation")
-
 		msm, err := st.mutator(adt.AsStore(rt)).withDealStates(WritePermission).
 			withPendingProposals(ReadOnlyPermission).withDealProposals(ReadOnlyPermission).build()
 		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load state")
 
-		for _, dealID := range params.DealIDs {
-			// This construction could be replaced with a single "update deal state" state method, possibly batched
-			// over all deal ids at once.
-			_, found, err := msm.dealStates.Get(dealID)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to get state for dealId %d", dealID)
-			if found {
-				rt.Abortf(exitcode.ErrIllegalArgument, "deal %d already included in another sector", dealID)
+		for _, req := range params.Requests {
+			_, _, err := ValidateDealsForActivation(&st, store, req.DealIDs, minerAddr, req.SectorExpiry, currEpoch)
```

Comment:
We're loading state here after we start mutating. This isn't going to be correct. I think we need to inline the logic from this function.